File: GuardianMonsters/inventory/src/main/java/de/limbusdev/guardianmonsters/inventory/AbilityChoiceSubMenu.java
package main.java.de.limbusdev.guardianmonsters.inventory;

import com.badlogic.ashley.signals.Signal;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton;
import com.badlogic.gdx.scenes.scene2d.ui.ScrollPane;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.ui.VerticalGroup;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.ArrayMap;
import com.badlogic.gdx.utils.I18NBundle;

import java.util.Observable;
import java.util.Observer;

import de.limbusdev.guardianmonsters.guardians.Element;
import de.limbusdev.guardianmonsters.guardians.GuardiansServiceLocator;
import de.limbusdev.guardianmonsters.guardians.abilities.Ability;
import de.limbusdev.guardianmonsters.guardians.abilities.AbilityGraph;
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;
import de.limbusdev.guardianmonsters.services.Services;
import de.limbusdev.guardianmonsters.ui.Constant;
import de.limbusdev.guardianmonsters.ui.widgets.Callback;
import de.limbusdev.guardianmonsters.ui.widgets.HoneyComb7ButtonsWidget;
import de.limbusdev.guardianmonsters.ui.widgets.SimpleClickListener;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.abilities.AbilityDetailWidget;
import main.java.de.limbusdev.guardianmonsters.inventory.ui.widgets.team.TeamMemberSwitcher;

/**
 * AbilityChoiceSubMenu
 *
 * @author Georg Eckert 2017
 */

public class AbilityChoiceSubMenu extends AInventorySubMenu
    implements TeamMemberSwitcher.Callbacks, AbilityDetailWidget.Callbacks, Callback.ButtonID,
        Observer
{

    private TeamMemberSwitcher switcher;
    private ArrayMap<Integer, AGuardian> team;
    private ButtonGroup<TextButton> abilityButtons;
    private AbilityDetailWidget details;
    private HoneyComb7ButtonsWidget abilitySlotButtons;
    private ImageButton back;
    private int currentlyChosenAbility;
    private Table abilityButtonTable;

    private Group abilitySlotChoice;

    public AbilityChoiceSubMenu(Skin skin, ArrayMap<Integer, AGuardian> team) {

        super(skin);
        this.team = team;

        currentlyChosenAbility = 0;

        switcher = new TeamMemberSwitcher(skin, team, this);
        details = new AbilityDetailWidget(skin, this, "button-switch");
        abilitySlotButtons = new HoneyComb7ButtonsWidget(Services.getUI().getBattleSkin(), this, HoneyComb7ButtonsWidget.ABILITY_ORDER);
        back = new ImageButton(skin, "button-back");

        back.addListener(new SimpleClickListener(() -> {
                abilitySlotChoice.remove();
        }));

        layout(skin);
        init(team.get(0), 0);
    }

    /**
     * Initializes the menu with the data of the given monster
     * @param m
     */
    private void init(AGuardian m, int teamPosition) {

        team.get(switcher.getCurrentlyChosen()).deleteObserver(this);

        switcher.init(m, teamPosition);

        refreshAbilitySlotButtons();

        if(m.getAbilityGraph().getLearntAbilities().size != 0) {

            currentlyChosenAbility = m.getAbilityGraph().getLearntAbilities().firstKey();
            showAbilityDetails();
        }

        m.addObserver(this);
    }

    // ................................................................ TeamMemberSwitcher.Callbacks
    @Override
    public void onChanged(int position) {

        propagateSelectedGuardian(position);
    }

    @Override
    public void refresh() {

    }

    @Override
    protected void layout(Skin skin) {

        abilityButtonTable = new Table();
        abilityButtonTable.align(Align.topLeft);

        switcher.setPosition(2, 202, Align.topLeft);
        addActor(switcher);

        ScrollPane scrollPane = new ScrollPane(abilityButtonTable, skin);
        scrollPane.setSize(140,204);
        scrollPane.setPosition(100,0);
        scrollPane.setScrollBarPositions(false,true);
        addActor(scrollPane);

        abilityButtons = new ButtonGroup<>();
        abilityButtons.setMaxCheckCount(1);
        abilityButtons.setMaxCheckCount(1);

        details.setPosition(Constant.WIDTH-2, 2, Align.bottomRight);
        addActor(details);

        abilitySlotChoice = new Group();
        abilitySlotChoice.setSize(Constant.WIDTH, Constant.HEIGHT);
        abilitySlotChoice.setPosition(0,0,Align.bottomLeft);
        Image overlay = new Image(getSkin().getDrawable("black-a80"));
        overlay.setSize(Constant.WIDTH, Constant.HEIGHT);
        overlay.setPosition(0,0,Align.bottomLeft);
        abilitySlotChoice.addActor(overlay);

        abilitySlotButtons.setPosition(0,32,Align.bottomLeft);
        abilitySlotChoice.addActor(abilitySlotButtons);

        back.setPosition(Constant.WIDTH-4,4,Align.bottomRight);
        abilitySlotChoice.addActor(back);
    }

    private void showAbilityDetails() {

        AGuardian guardian = team.get(switcher.getCurrentlyChosen());
        details.init(guardian, currentlyChosenAbility, true);
    }



    // ............................................................... AbilityDetailWidget.Callbacks
    @Override
    public void onLearn(int nodeID) {

        addActor(abilitySlotChoice);
        refreshAbilitySlotButtons();
    }

    private void refreshAbilitySlotButtons() {

        System.out.println("REfresh Buttons");

        abilityButtons.clear();
        abilityButtonTable.clearChildren();

        I18NBundle translation = Services.getL18N().Abilities();

        AGuardian m = team.get(switcher.getCurrentlyChosen());

        for(final int key : m.getAbilityGraph().getLearntAbilities().keys()) {

            Ability.aID abilityID = m.getAbilityGraph().getLearntAbilities().get(key);

            if(abilityID != null) {

                Ability ability = GuardiansServiceLocator.getAbilities().getAbility(abilityID);
                TextButton tb = new TextButton(translation.get(ability.name), getSkin(), "item-button-sandstone");

                tb.addListener(new SimpleClickListener(() -> {

                    currentlyChosenAbility = key;
                    showAbilityDetails();
                }));

                abilityButtonTable.row().spaceBottom(1);
                abilityButtonTable.add(tb).width(140).height(32);
                abilityButtons.add(tb);

                if (key == 0) {
                    tb.setChecked(true);
                }
            }
        }
    }

    // ................................................................ SevenButtonsWidget.Callbacks
    @Override
    public void onClick(int buttonID) {

        AGuardian guardian = team.get(switcher.getCurrentlyChosen());
        guardian.getAbilityGraph().setActiveAbility(buttonID, currentlyChosenAbility);
        refreshAbilitySlotButtons();
    }

    @Override
    public void syncSelectedGuardian(int teamPosition) {

        init(team.get(teamPosition), teamPosition);
    }

    @Override
    public void update(Observable observable, Object o) {

        refresh();
        refreshAbilitySlotButtons();
    }
}

File: GuardianMonsters/inventory/src/de/limbusdev/guardianmonsters/inventory/AbilityChoiceSubMenu.kt
package de.limbusdev.guardianmonsters.inventory

import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.utils.Align

import java.util.Observable
import java.util.Observer

import de.limbusdev.guardianmonsters.guardians.GuardiansServiceLocator
import de.limbusdev.guardianmonsters.guardians.abilities.Ability
import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian
import de.limbusdev.guardianmonsters.guardians.monsters.Team
import de.limbusdev.guardianmonsters.inventory.ui.widgets.abilities.AbilityDetailWidget
import de.limbusdev.guardianmonsters.inventory.ui.widgets.team.TeamMemberSwitcher
import de.limbusdev.guardianmonsters.scene2d.*
import de.limbusdev.guardianmonsters.services.Services
import de.limbusdev.guardianmonsters.ui.Constant
import de.limbusdev.guardianmonsters.ui.widgets.Callback
import de.limbusdev.guardianmonsters.ui.widgets.HoneyComb7ButtonsWidget
import de.limbusdev.guardianmonsters.ui.widgets.HoneyComb7ButtonsWidget.ABILITY_ORDER
import de.limbusdev.utils.logInfo
import ktx.style.get

/**
 * AbilityChoiceSubMenu
 *
 * @author Georg Eckert 2017
 */

class AbilityChoiceSubMenu
(
        private val skin: Skin,
        private val team: Team
)
    : AInventorySubMenu(), TeamMemberSwitcher.Callbacks, AbilityDetailWidget.Callbacks, Callback.ButtonID, Observer
{
    // --------------------------------------------------------------------------------------------- PROPERTIES
    companion object { const val TAG = "AbilityChoiceSubMenu" }

    private val switcher            : TeamMemberSwitcher
    private var abilityButtons      : ButtonGroup<TextButton> = ButtonGroup()
    private val details             : AbilityDetailWidget
    private val abilitySlotButtons  : HoneyComb7ButtonsWidget
    private val back                : ImageButton
    private var abilityButtonTable  : Table = Table()
    private var abilitySlotChoice   : Group = Group()

    private var currentlyChosenAbility: Int = 0


    // --------------------------------------------------------------------------------------------- CONSTRUCTORS
    init
    {
        currentlyChosenAbility = 0

        switcher = TeamMemberSwitcher(skin, team, this)
        details = AbilityDetailWidget(skin, this, "button-switch")
        abilitySlotButtons = HoneyComb7ButtonsWidget(Services.UI().battleSkin, this, ABILITY_ORDER)

        back = ImageButton(skin, "button-back")
        back.replaceOnButtonClick { abilitySlotChoice.remove() }

        layout(skin)
        initialize(team[0], 0)
    }

    /**
     * Initializes the menu with the data of the given monster
     * @param guardian
     */
    private fun initialize(guardian: AGuardian, teamPosition: Int)
    {
        team[switcher.currentlyChosen].deleteObserver(this)

        switcher.init(guardian, teamPosition)

        refreshAbilitySlotButtons()

        if (guardian.abilityGraph.learntAbilities.size > 0)
        {
            currentlyChosenAbility = guardian.abilityGraph.learntAbilities.firstKey()
            showAbilityDetails()
        }

        guardian.addObserver(this)

        reloadHoneyCombWidget(guardian)

        abilitySlotButtons.setCallbacks { buttonID ->

            guardian.abilityGraph.setActiveAbility(buttonID, currentlyChosenAbility)
            reloadHoneyCombWidget(guardian)
        }
    }

    private fun reloadHoneyCombWidget(guardian: AGuardian)
    {
        // Initialize HoneyComb Button Widget
        for(i in 0..6) { abilitySlotButtons.setButtonText(i, "") }
        for(ability in guardian.abilityGraph.activeAbilities)
        {
            if(ability.value != null)
            {
                abilitySlotButtons.setButtonText(ability.key, GuardiansServiceLocator.abilities[ability.value])
            }
        }
    }


    // --------------------------------------------------------------------------------------------- CALLBACKS
    // ................................................................ TeamMemberSwitcher.Callbacks
    override fun onChanged(position: Int) = propagateSelectedGuardian(position)

    override fun refresh() {}

    override fun layout(skin: Skin)
    {
        abilityButtonTable.align(Align.topLeft)

        switcher.setPosition(2f, 202f, Align.topLeft)
        addActor(switcher)

        // Put ability buttons into a scrollable panel
        val scrollPane = makeScrollPane(abilityButtonTable, skin, Scene2DLayout(140f, 204f, 100f, 0f), false, true, this)

        // Exactly one button must be checked at all times
        abilityButtons.setMaxCheckCount(1)
        abilityButtons.setMaxCheckCount(1)

        details.setPosition(Constant.WIDTHf - 2, 2f, Align.bottomRight)
        addActor(details)

        abilitySlotChoice.setSize(Constant.WIDTHf, Constant.HEIGHTf)
        abilitySlotChoice.setPosition(0f, 0f, Align.bottomLeft)

        // Setup transparent black overlay
        val overlay = makeImage(skin["black-a80"], ImgLayout(Constant.WIDTHf, Constant.HEIGHTf), abilitySlotChoice)

        abilitySlotButtons.setPosition(0f, 32f, Align.bottomLeft)
        abilitySlotChoice.addActor(abilitySlotButtons)

        back.setPosition(Constant.WIDTHf, 4f, Align.bottomRight)
        abilitySlotChoice.addActor(back)
    }

    private fun showAbilityDetails()
    {
        val guardian = team[switcher.currentlyChosen]
        details.init(guardian, currentlyChosenAbility, true)
    }


    // ............................................................... AbilityDetailWidget.Callbacks
    override fun onLearn(nodeID: Int)
    {
        addActor(abilitySlotChoice)
        refreshAbilitySlotButtons()
    }

    private fun refreshAbilitySlotButtons()
    {
        logInfo(TAG) { "Refresh Buttons" }

        abilityButtons.clear()
        abilityButtonTable.clearChildren()

        val i18n = Services.I18N().Abilities()

        val guardian = team[switcher.currentlyChosen]

        for (key in guardian.abilityGraph.learntAbilities.keys())
        {
            val abilityID = guardian.abilityGraph.learntAbilities.get(key)

            if (abilityID != null)
            {
                val ability = GuardiansServiceLocator.abilities.getAbility(abilityID)
                val button = TextButton(i18n.get(ability.name), skin, "item-button-sandstone")

                button.replaceOnButtonClick {

                    currentlyChosenAbility = key
                    showAbilityDetails()
                }

                abilityButtonTable.row().spaceBottom(1f)
                abilityButtonTable.add(button).width(140f).height(32f)
                abilityButtons.add(button)

                if (key == 0) { button.isChecked = true }
            }
        }
    }

    // ................................................................ SevenButtonsWidget.Callbacks
    override fun onClick(buttonID: Int)
    {
        val guardian = team[switcher.currentlyChosen]
        guardian.abilityGraph.setActiveAbility(buttonID, currentlyChosenAbility)
        refreshAbilitySlotButtons()
    }

    override fun syncSelectedGuardian(teamPosition: Int)
    {
        initialize(team[teamPosition], teamPosition)
    }

    override fun update(observable: Observable, o: Any?)
    {
        refresh()
        refreshAbilitySlotButtons()
    }
}
